### PR TITLE
Convert Guided tour data from local storage to user settings

### PR DESCRIPTION
**Pull request by sahil143**
**Fixes**: https://issues.redhat.com/browse/ODC-5083

**Analysis / Root cause**: convert the local storage to user settings

**Solution Description**: use `useUserSettings` hook to store the local storage data into a config map.

**Screen shots / Gifs for design review**: No UI changes

**Unit test coverage report**: Added

**Test setup:**

**Browser conformance**: 

- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

### CODE REVIEW

File: frontend/packages/console-app/src/components/tour/tour-context.ts [hunk 1]
```diff
@@ -81,6 +82,43 @@ const getRequiredFlagsByTour = createSelector(
...
+
+const getTourLocalStorageData = (): TourLocalStorageData =>
+  JSON.parse(localStorage.getItem(TOUR_LOCAL_STORAGE_KEY));
+
+const useTourStateForPerspective = (
+  perspective: string,
+): [TourLocalStorageType, (completed: boolean) => void] => {
+  const [tourLocalState, setTourLocalState] = useState<TourLocalStorageData>(
+    getTourLocalStorageData() ?? {},
+  );
```
Comment (sahil143) [inline frontend/packages/console-app/src/components/tour/tour-context.ts:101]: Replace this `useState` hook with `useUserSettings` hook

File: frontend/packages/console-app/src/components/tour/tour-context.ts [hunk 2]
```diff
@@ -16,11 +17,12 @@ import { getFlagsObject } from '@console/internal/reducers/features';
 import { useExtensions } from '@console/plugin-sdk';
 import { RootState } from '@console/internal/redux';
 import { isGuidedTour, GuidedTour } from '@console/plugin-sdk/src/typings/guided-tour';
-import { TourActions } from './const';
+import { useUserSettingsCompatibility } from '@console/shared/src/hooks/useUserSettingsCompatibility';
+import { TourActions, TOUR_LOCAL_STORAGE_KEY } from './const';
 import {
-  getTourLocalStorageForPerspective,
+  // getTourLocalStorageForPerspective,
```
Comment (invincibleJai) [inline frontend/packages/console-app/src/components/tour/tour-context.ts:23]: nit: can remove comment

File: frontend/packages/console-app/src/components/tour/tour-context.ts [hunk 3]
```diff
@@ -81,6 +83,43 @@ const getRequiredFlagsByTour = createSelector(
...
 
+type TourLocalStorageType = {
+  completed: boolean;
+};
+
+type TourLocalStorageData = {
+  [key: string]: TourLocalStorageType;
+};
+
+const TOUR_CONFIGMAP_KEY = `console.guidedTour`;
```
Comment (invincibleJai) [inline frontend/packages/console-app/src/components/tour/tour-context.ts:94]: maybe can store `console` as const here console-shared/src/constants/common.ts , can be used for all scenarios. WDYT?

File: frontend/packages/console-app/src/components/tour/tour-context.ts [hunk 4]
```diff
@@ -107,12 +149,13 @@ export const useTourValuesForContext = (): TourContextType => {
...
 
   useEffect(() => {
-    const { completed: initCompleted } = getTourLocalStorageForPerspective(activePerspective);
-    tourDispatch({ type: TourActions.initialize, payload: { completed: initCompleted } });
+    tourDispatch({ type: TourActions.initialize, payload: { completed } });
     setPerspective(activePerspective);
-  }, [activePerspective]);
+    // only run effect when the active perspective changes
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [activePerspective, loaded]);
```
Comment (invincibleJai) [inline frontend/packages/console-app/src/components/tour/tour-context.ts:156]: is `loaded` intentional here as only used as dependency ?

File: frontend/packages/console-app/src/components/tour/tour-context.ts (hunk 3 above)
Comment (sahil143) [reply frontend/packages/console-app/src/components/tour/tour-context.ts:94]: It's fine to keep these in their respective files as they are not being used anywhere else.

File: frontend/packages/console-app/src/components/tour/tour-context.ts (hunk 4 above)
Comment (sahil143) [reply frontend/packages/console-app/src/components/tour/tour-context.ts:156]: Yes, loaded is intentional. If state doesn't exist for a perspective after the config map is loaded, this effect sets the default state for that perspective

**Comment by invincibleJai** [comment]
Thanks @sahil143 looks good , we'll revisit this later once we have required changes in

**Comment by invincibleJai** [comment]
/lgtm

Verified the changes , works as expected

**Comment by jerolimov** [comment]
Tested this locally, works fine.
/lgtm

@sahil143 @invincibleJai Noticed one little strange thing here, because the component and hook was never unmounted, it does not removes the local storage key. But when removing the entry from the ConfigMap the old localStorage key is still available.
It's an edge case, but maybe we should remove the localStorage keys when the ConfigMap was loaded or updated successfully instead of unmount? Wdyt?

/assign @christianvogt
Has lgtm from Sahil, Jai and me. Can you check it and give an approval if everything is fine? Thanks

**Comment by christianvogt** [comment]
@invincibleJai @jerolimov maybe we go back to also removing the key when we set a new value

**Pull request merged**